### PR TITLE
[Breaking] Standardize Trajectory Transformation Logic Across Backends

### PR DESCRIPTION
**Pull request by listar2000**
## Overview

```
┌─────────────────────────────────────────────────────────────────────────────┐
│                           list[Episode]                                     │
│                   (Backend-agnostic, shared)                                │
└─────────────────────────────────────────────────────────────────────────────┘
                    │
                    ▼
┌─────────────────────────────────────────────────────────────────────────────┐
│                        list[TrajectoryGroup]                                │
│           (Backend-agnostic, used for advantage computation)                │
└─────────────────────────────────────────────────────────────────────────────┘
                    │
        ┌───────────┴───────────┐
        ▼                       ▼
┌───────────────────┐   ┌───────────────────┐
│    DataProto      │   │  list[Datum]      │
│      (verl)       │   │    (Tinker)       │
│                   │   │                   │
│ - Tensor batches  │   │ - Individual      │
│ - Non-tensor meta │   │   training        │
│ - World-size      │   │   examples        │
│   aligned         │   │ - Numpy/Tensor    │
└───────────────────┘   └───────────────────┘
```

---

This PR standardizes the transformation logic from `Episode` → `TrajectoryGroup` → backend-specific training data (e.g., `DataProto` for verl, `Datum` for Tinker). The goal is to enable downstream unification of trainers by 

1. Extracting **backend-agnostic** components into a shared `rllm/trainer/common/` module.
2. Porting all `TrajectoryGroup` → **backend-specific** data structure code into the `rllm/trainer/[backend]/[backend]_data_processor/` modules, where `backend = verl/tinker` for now.

**Theme**: Support for unified trainer abstraction by decoupling the training loop stages into shared and backend-specific components.

---

## New Module: `rllm/trainer/common/`

A new module has been created to house utilities shared across all trainer backends.

### `config.py` — Shared Configuration Dataclasses

Defines four configuration dataclasses used by both trainers:

- **`CompactFilteringConfig`**: Controls filtering of episodes based on termination reasons (e.g., timeout, max turns exceeded, errors). Provides a `should_mask()` method that returns `True` if an episode with a given termination reason should be excluded from training.

- **`TransformConfig`**: Controls the episode-to-trajectory-group transformation pipeline:
  - `impute_missing_names`: Auto-rename unnamed trajectories (e.g., `agent_0`, `agent_1`)
  - `drop_unnamed_traj`: Whether to drop trajectories without names
  - `broadcast`: Whether to use trajectory-level rewards (broadcast) or per-step rewards

- **`RejectionSamplingConfig`**: Controls rejection sampling behavior:
  - `mode`: `"none"`, `"episode"`, or `"group"` (group mode not yet implemented)
  - `min_trajs_per_group`: Minimum trajectories required per group
  - `min_partial_solve_tasks`: Minimum tasks with partial solves before proceeding (for episode mode)

- **`AlgorithmConfig`**: Algorithm parameters for advantage computation:
  - `estimator`: `rLLMAdvantageEstimator.GRPO` or `REINFORCE`
  - `stepwise_advantage_mode`: `"broadcast"` or `"per_step"`
  - `norm_adv_by_std_in_grpo`: Whether to normalize advantages by std in GRPO

### `transform.py` — Episode → TrajectoryGroup Pipeline

Provides the main transformation pipeline via `transform_episodes_to_trajectory_groups()`:

1. **`impute_trajectory_names()`**: Renames trajectories that have the default name to position-based names (`default_traj_name_0`, etc.). Can optionally drop unnamed trajectories.

2. **`build_trajectory_groups()`**: Groups trajectories by `{task_id}:{trajectory_name}`. This means all trajectories of the same type (e.g., "solver") from the same task are grouped together for advantage computation. Applies compact filtering here — episodes with masked termination reasons are skipped.

3. **`validate_and_propagate_rewards()`**: Ensures reward consistency:
   - In `broadcast` mode: Propagates last-step reward to trajectory-level if missing
   - In `per_step` mode: Validates all trajectories in a group have the same step count

Returns both the trajectory groups and grouping metrics (num groups, avg/min/max group size).

### `rejection_sampling.py` — Rejection Sampling Logic

Provides rejection sampling that works with `TrajectoryGroup`:

- **`RejectionSamplingMetrics`**: Tracks `solve_none`, `solve_all`, `solve_partial` counts (episodes where all/none/some rollouts are correct), plus group filtering counts.

- **`RejectionSamplingState`**: Maintains state across batches for episode-level accumulation mode.

- **`filter_groups()`**: Filters groups that don't meet `min_trajs_per_group` threshold.

- **`filter_episodes()`**: Removes trajectories from episodes that correspond to dropped groups (using trajectory UIDs).

- **`apply_rejection_sampling_and_filtering()`**: Main entry point. In `"episode"` mode, accumulates groups until enough partial solves exist; in `"none"` mode, just returns filtered groups.

### `advantage.py` — TrajectoryGroup-based Advantage Computation

Provides `compute_advantage_from_trajectory_groups()`:

- For `broadcast` mode: Computes GRPO/REINFORCE advantages using trajectory-level rewards, then broadcasts the advantage to all steps within each trajectory.
- For `per_step` mode: Computes advantages step-by-step across trajectories in the group.

### `metrics.py` — Common Metrics

Provides `reduce_reward_metrics_by_trajectory_name()` which computes mean/max/min/std reward statistics per trajectory name across all groups.

---

### Major Refactoring

- **No longer inherits from `TinkerAgentTrainer`**: Now a standalone class with its own full training loop implementation in `_fit_agent_async()`.

- **Uses shared common utilities**: Now imports and uses `CompactFilteringConfig`, `TransformConfig`, `RejectionSamplingConfig`, `AlgorithmConfig`, `build_trajectory_groups`, and `reduce_reward_metrics_by_trajectory_name` from `rllm.trainer.common`.

- **New `_validate_and_setup_configs()` method**: Sets up all four configuration objects during initialization.

- **Training loop (`_fit_agent_async`)**: Fully rewritten to:
  - Use `build_interleave_batch()` for batch preparation with UIDs
  - Call `generate_episodes()` (async generator) to produce episodes
  - Delegate training step logic to `TinkerPolicyTrainer.step()` which now uses the common transform pipeline
  - Use `visualize_trajectory_last_steps()` from the new utils module
  - Use `compute_training_metrics()` and `print_metrics_table()` for metrics logging

- **Validation flow**: `validate_agent()` now uses `build_trajectory_groups()` and `reduce_reward_metrics_by_trajectory_name()` for computing validation metrics.

### `TinkerPolicyTrainer` Changes

- Now accepts optional `cf_config`, `transform_config`, and `algorithm_config` parameters.
- `step()` method now:
  1. Calls `transform_episodes_to_trajectory_groups()` from common module
  2. Calls `transform_trajectory_groups_to_datums()` from the new `tinker_data_processor` module
  3. Returns `(datums, trajectory_groups, logprobs, grouping_metrics)` instead of the previous signature
- Added `@require_training_client` decorator for methods that need the training client.

### New `tinker_data_processor/transform.py`

- Provides `build_datum_from_step()` and `transform_trajectory_groups_to_datums()`.
- Internally calls `compute_advantage_from_trajectory_groups()` from the common module before building Tinker `Datum` objects.

---

### Major Refactoring

- **Uses shared common utilities**: Now imports `CompactFilteringConfig`, `TransformConfig`, `RejectionSamplingConfig`, `AlgorithmConfig`, `transform_episodes_to_trajectory_groups`, `apply_rejection_sampling_and_filtering`, and `compute_advantage_from_trajectory_groups` from `rllm.trainer.common`.

- **New `_validate_and_setup_configs()` method**: Replaces the old `_validate_config()`. Sets up all four shared configuration objects.

- **Renamed method**: `generate_trajectories()` → `generate_episodes()` to align with the Episode-centric data flow. Now returns `list[Episode]` instead of `DataProto`.

- **New helper methods extracted from `fit_agent()`**:
  - `_update_episode_metrics_and_termination_counts()`: Collects workflow metrics and termination reasons from episodes
  - `_compute_step_level_values()`: Computes old_log_probs, ref_log_probs, and critic values
  - `_update_policy()`: Updates actor and critic networks
  - `_compute_advantage_verl()`: The original verl-native advantage computation (retained as fallback)

- **Training loop (`fit_agent`)** refactored to:
  1. Generate episodes via `generate_episodes()`
  2. Transform to trajectory groups via `transform_episodes_to_trajectory_groups()`
  3. Apply rejection sampling via `apply_rejection_sampling_and_filtering()` with new `RejectionSamplingState`
  4. Compute advantages via `compute_advantage_from_trajectory_groups()` (when `use_rllm_advantage=True`)
  5. Transform to `DataProto` via new `transform_episodes_to_dataproto()`
  6. Continue with step-level values and policy update

- **Validation loop (`_validate_agent`)**: Now uses `generate_episodes()` and `transform_episodes_to_dataproto()`.

- **Removed inline visualization**: The old `visualize_trajectory_last_step()` method (130+ lines) has been removed and replaced with `visualize_trajectory_last_steps()` from the shared utils.

### New `verl_data_processor/` Module

- **`dataclass.py`**: Defines `ProcessedStepData` and `AccumulatedData` for accumulating step data during transformation.

- **`transform.py`**: Provides `transform_episodes_to_dataproto()` and `transform_trajectory_groups_to_dataproto()`:
  - Handles prompt/response padding
  - Builds attention masks and position IDs (including multimodal support from #315)
  - Constructs reward tensors at the correct positions
  - Includes pre-computed advantages from the common advantage computation if available

---

## Changes to `AgentWorkflowEngine`

- **Simplified `execute_tasks_verl()`**: Now returns `list[Episode]` directly instead of `DataProto`. The transformation to `DataProto` is now done in the trainer.

_[Eventually the goal is to *remove backend-dependent functions like execute_tasks_verl()* completely]_

- **Removed `transform_results_for_verl()`**: This ~250 line method has been moved to the new `verl_data_processor/transform.py` module.

- **Removed `_handle_multimodal_position_ids()`**: Moved to the new transform module.

---

## Changes to `rllm/agents/agent.py`

- **New `TrajectoryGroup` dataclass**: A container for trajectories that should be compared for advantage computation. Unlike `Episode` (raw rollout data), `TrajectoryGroup` is structured for the advantage computation stage.

- **`Step.advantage` field**: Added to store the computed advantage value.

- **`Step.__post_init__()` method**: Added to automatically backfill `prompt_ids`, `response_ids`, and `logprobs` from `model_output` if present.

- **`Trajectory.reward` type change**: Changed from `float = 0.0` to `float | None = None` to explicitly represent cases where trajectory-level reward doesn't exist (will be propagated from last step).

- **`_DEFAULT_TRAJ_NAME` constant**: Exported for use in configuration.

---

### `visualization.py`

- `visualize_trajectory_last_steps()`: Unified visualization function used by both trainers. Samples trajectory groups, picks a random trajectory from each, and displays the last step with colored tokens.
- `VisualizationConfig`: Dataclass for customizing colors/styles.
- `colorful_print()`, `colorful_warning()`: Click-based colored output helpers (moved from `rllm/misc.py`).
- `abbreviate_string()`: Truncates long strings while preserving beginning and end.

### `performance.py`

- `marked_timer()`, `simple_timer()`: Context managers for timing code blocks (adapted from verl's profiler).

### `tracking.py`

- `Tracking`: A wrapper class for logging to various backends (console, wandb, tensorboard).

---

### `agent_ppo_trainer.yaml`

- Added `min_trajs_per_group` option under `rllm.rejection_sample`.

### `tinker_rl_trainer.yaml`

- Added full `rllm` config section with `compact_filtering`, `rejection_sample`, and `stepwise_advantage` to match verl config structure.

---

## New SDK Example

A new example has been added under `examples/sdk/dispatch_search/`.

---

## Open TODOs

The following TODOs have been added in this PR for future work:

### Configuration & Hydra Integration

| File | TODO |
|------|------|
| `trainer/verl/agent_workflow_trainer.py` | Add the four configuration dataclasses (`CompactFilteringConfig`, `TransformConfig`, etc.) to the Hydra config system for proper YAML configuration. |

### Feature Gaps

| File | TODO |
|------|------|
| `trainer/common/config.py` | Add more advantage estimators to `rLLMAdvantageEstimator` (currently only GRPO and REINFORCE). |
| `trainer/common/rejection_sampling.py` | Implement group-level rejection sampling mode (`mode="group"`). |
| `trainer/common/advantage.py` | Support per-trajectory-group advantage modes (different estimators/settings for different trajectory types). |
| `trainer/tinker/tinker_workflow_trainer.py` | Implement mini-batching support. Currently only `num_minibatches=1` is supported. |
| `trainer/verl/verl_data_processor/transform.py` | Support `token_level_scores` from the `Step` attribute level, and implement the `kl_penalty` logic used in verl. |

### Metrics & Logging

| File | TODO |
|------|------|
| `trainer/common/metrics.py` | Think hard about what are the actually important metrics in agentic RL settings. |
| `trainer/tinker/tinker_metrics_utils.py` | Implement separate episode-based metrics (currently trajectory-group-based). |

### Code Cleanup

| File | TODO |
|------|------|
| `trainer/tinker/tinker_metrics_utils.py` | Delete or refactor the `print_episodes()` function (used by `tinker_agent_trainer`). |
| `engine/agent_workflow_engine.py` | Refactor `execute_tasks()` to remove the `task_states` dictionary pattern. |

---

## Breaking Changes & Side Effects

As the unification is still on-going, the changes implemented so far will likely caused side-effects and errors for:
-  **Agent-env trainers** (e.g. `agent_ppo_trainer.py` and `tinker_agent_trainer`)
- Trainers directly **adapted from older workflow trainers** (e.g. `agent_workflow_trainer_fireworks.py` and `agent_sdk_trainer.py`)
- Cumulative agent's behavior might cause some weird issue for now (need to have a special module to turn cumulative agent into a regular workflow trajectory)

### CODE REVIEW

**Comment by listar2000** [comment]
## Planning towards full unification

Changes introduced in this PR enable us to have **almost identical processing pipeline** in the two trainers. Namely:

###  Training Loop Comparison

| Stage | Tinker | verl | Shareable? |
|-------|--------|------|------------|
| **Initialization** | `__init__` + `_validate_and_setup_configs` | `__init__` + `_validate_and_setup_configs` + `init_workers` | Partial (configs yes, workers no) |
| **Load/Resume Checkpoint** | `trainer.initialize_async` | `_load_checkpoint` | No (backend-specific) |
| **Pre-validation** | `validate_agent` | `_validate_agent` | Yes (structure similar) |
| **Batch Preparation** | `build_interleave_batch` | `DataProto.repeat` + uuid assignment | Partial |
| **Episode Generation** | `generate_episodes` → `AgentWorkflowEngine` | `generate_episodes` → `AgentWorkflowEngine` | **Yes** |
| **Transform to Groups** | `transform_episodes_to_trajectory_groups` | `transform_episodes_to_trajectory_groups` | **Yes** |
| **Rejection Sampling** | `RejectionSamplingConfig` (unused currently) | `apply_rejection_sampling_and_filtering` | **Yes** |
| **Advantage Computation** | `compute_advantage_from_trajectory_groups` | `compute_advantage_from_trajectory_groups` (or verl native) | **Yes** |
| **Transform to Training Data** | `transform_trajectory_groups_to_datums` | `transform_episodes_to_dataproto` | No (different data structures) |
| **Compute Step-Level Values (e.g. `old_log_prob`)** | (included in `trainer.step`, handled by `Tinker`) | `_compute_step_level_values` | No |
| **Update Policy** | `trainer.step` + `optim_step_future` | `_update_policy` | No |
| **Compute Metrics** | `compute_training_metrics` | `verl` metric utils | Partial |
| **Log Metrics** | `Tracking.log` | `Tracking.log` | **Yes** |
| **Save Checkpoint** | `trainer.save_checkpoint_async` | `_save_checkpoint` | No |
| **Validation** | `validate_agent` | `_validate_agent` | Partial |

### Shared Components Already Unified

From `rllm/trainer/common`:
- `CompactFilteringConfig` - filtering based on termination reasons
- `TransformConfig` - episode-to-group transformation settings
- `RejectionSamplingConfig` - rejection sampling parameters
- `AlgorithmConfig` - advantage estimator settings (GRPO, REINFORCE)
- `transform_episodes_to_trajectory_groups` - transformation pipeline
- `compute_advantage_from_trajectory_groups` - advantage computation
- `apply_rejection_sampling_and_filtering` - rejection sampling logic

**Comment by listar2000** [comment]
An example design (by Claude, for reference only) of the new unified trainer interface that can be built on top of this PR is at https://hackmd.io/@listar2000/BypOdPGzWg/edit

File: rllm/agents/agent.py [hunk 1]
```diff
@@ -16,14 +21,32 @@ class Step:
...
+    model_output: ModelOutput | None = None
     info: dict = field(default_factory=dict)  # Store any additional info.
 
     # field below are filled by the engine
     reward: float = 0.0
     done: bool = False
     mc_return: float = 0.0
 
+    # field below are filled by the advantage computer
+    advantage: float | None = None
```
Comment (listar2000) [inline rllm/agents/agent.py:33]: Eventually we might want to make `advantage: list[float]` as well to support token-level advantage instead of sequence level (e.g. DAPO).
